Added group norm L0 and shifted group norm L0

AHsu98 opened the pull request:
First commit towards adding a group L0 norm. Happy to change the name, perhaps L02 is better. This is defined as the L0 norm of the vector of L2 norms of the indices, weighted by the values of lambda. Still needs a bit of testing I think, but pretty simple, only a couple of lines to change from GroupL2.

Also, in both this and GroupL2, we should call out that we require it to be a separable sum (no indices overlap).

dpo commented on src/groupNormL0.jl:
```diff
@@ -0,0 +1,68 @@
+  idx::I
+
+  function GroupNormL0{R, RR, I}(lambda::RR, idx::I) where {R <: Real, RR <: AbstractVector{R}, I}
+    if any(lambda .< 0)
+      error("weights λ must be nonnegative")
+    elseif length(lambda) != length(idx)
+      error("number of weights and groups must be the same")
+    else
+      new{R, RR, I}(lambda, idx)
+    end
```
```suggestion
    any(lambda .< 0) && error("weights λ must be nonnegative")
    length(lambda) != length(idx) && error("number of weights and groups must be the same")
    new{R, RR, I}(lambda, idx)
```

dpo commented on src/groupNormL0.jl:
```diff
@@ -0,0 +1,68 @@
+  sum_c = R(0)
+  for (idx, λ) ∈ zip(f.idx, f.lambda)
+    y = norm(x[idx])
+    if y>0
```
```suggestion
    if y > 0
```

dpo commented on src/shiftedGroupNormL0.jl:
```diff
@@ -0,0 +1,77 @@
+  V1 <: AbstractVector{R},
+  V2 <: AbstractVector{R},
+}
+  ψ.sol .= q + ψ.xk + ψ.sj
```
```suggestion
  ψ.sol .= q .+ ψ.xk .+ ψ.sj
```

dpo commented on src/shiftedGroupNormL0.jl:
```diff
@@ -0,0 +1,77 @@
+      y[idx] .= ψ.sol[idx]
+    end
+  end
+  y .-= (ψ.xk + ψ.sj)
```
```suggestion
  y .-= (ψ.xk .+ ψ.sj)
```

dpo reviewed (changes requested):
Hey @AHsu98! Sorry I must have missed this PR. It looks great, thank you! Here are a few minor comments.

Would you mind adding some using tests though? Thanks!

MohamedLaghdafHABIBOULLAH commented:
Thank you @AHsu98, could you please rebase and update this PR ? 
It would be nice to add some simple tests and check for allocation using the macro  [```wrappedallocs```](https://github.com/JuliaSmoothOptimizers/ShiftedProximalOperators.jl/blob/master/test/test_allocs.jl#L28)

AHsu98 commented:
Alright, I rebased, made some updates (fixed some original mistakes), and incorporated the previous suggestions. It'd be nice to add a check that the indices are non-overlapping (the naive thing I tried worked for lists of integers as indices, but caused errors in the tests, maybe from shifts, or multiple indices). I also made the suggested changes to the checks on groupNormL2 as well. 
Still remains to add testing and check allocation.

dpo commented on src/groupNormL0.jl:
```diff
@@ -0,0 +1,64 @@
+  ysum = R(0)
+  for (idx, λ) ∈ zip(f.idx, f.lambda)
+    yt = norm(x[idx])^2
+    if yt !=0
```
```suggestion
    if yt > 0
```

dpo commented on src/groupNormL0.jl:
```diff
@@ -0,0 +1,64 @@
+for groups ``x_{[i]}`` and nonnegative weights ``\lambda_i``.
+This assumes that the groups ``x_{[i]}`` are non-overlapping
+"""
+struct GroupNormL0{R <: Real, RR <: AbstractVector{R}, I}
+  lambda::RR
+  idx::I
+
+  function GroupNormL0{R, RR, I}(lambda::RR, idx::I) where {R <: Real, RR <: AbstractVector{R}, I}
+    any(lambda .< 0) && error("weights λ must be nonnegative")
+    length(lambda) != length(idx) && error("number of weights and groups must be the same")
+    new{R, RR, I}(lambda, idx)
+  end
+end
```
```suggestion
struct GroupNormL0{R <: Real, V <: AbstractVector{R}, I}
  lambda::V
  idx::I

  function GroupNormL0{R, V, I}(lambda::V, idx::I) where {R <: Real, V <: AbstractVector{R}, I}
    any(lambda .< 0) && error("weights λ must be nonnegative")
    length(lambda) != length(idx) && error("number of weights and groups must be the same")
    new{R, V, I}(lambda, idx)
  end
end
```

dpo commented on src/groupNormL0.jl:
```diff
@@ -0,0 +1,64 @@
+  return sum_c
+end
+
+function prox!(
+  y::AbstractArray{R},
+  f::GroupNormL0{R, RR, I},
+  x::AbstractArray{R},
+  γ::R = R(1),
+) where {R <: Real, RR <: AbstractVector{R}, I}
```
```suggestion
function prox!(
  y::AbstractArray{R},
  f::GroupNormL0{R, V, I},
  x::AbstractArray{R},
  γ::R = R(1),
) where {R <: Real, V <: AbstractVector{R}, I}
```

dpo commented on src/groupNormL2.jl:
```diff
@@ -17,13 +17,9 @@ struct GroupNormL2{R <: Real, RR <: AbstractVector{R}, I}
   idx::I
 
   function GroupNormL2{R, RR, I}(lambda::RR, idx::I) where {R <: Real, RR <: AbstractVector{R}, I}
```
```suggestion
  function GroupNormL2{R, V, I}(lambda::V, idx::I) where {R <: Real, V <: AbstractVector{R}, I}
```

dpo commented on src/groupNormL2.jl:
```diff
@@ -17,13 +17,9 @@ struct GroupNormL2{R <: Real, RR <: AbstractVector{R}, I}
-    end
+    any(lambda .< 0) && error("weights λ must be nonnegative")
+    length(lambda) != length(idx) && error("number of weights and groups must be the same")
+    new{R, RR, I}(lambda, idx)
```
```suggestion
    new{R, V, I}(lambda, idx)
```

dpo replied:
There may be others in this file. It’s just easier to remember that V stands for “vector”.

dpo commented on src/shiftedGroupNormL0.jl:
```diff
@@ -0,0 +1,77 @@
+
+mutable struct ShiftedGroupNormL0{
+  R <: Real,
+  RR <: AbstractVector{R},
```
Use V0, V1, V2, V3 here.
